src/parser/MagicServiceQuery.cpp

```diff
@@ -0,0 +1,55 @@
+// Copyright 2024, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors: Johannes Herrmann <johannes.r.herrmann(at)gmail.com>
+//          Christoph Ullinger <ullingec@informatik.uni-freiburg.de>
+
+#include "parser/MagicServiceQuery.h"
+
+#include "parser/GraphPatternOperation.h"
+
+namespace parsedQuery {
+
+// ____________________________________________________________________________
+void MagicServiceQuery::addBasicPattern(const BasicGraphPattern& pattern) {
+  for (SparqlTriple triple : pattern._triples) {
+    addParameter(triple);
+  }
+}
+
+// ____________________________________________________________________________
+void MagicServiceQuery::addGraph(const GraphPatternOperation& op) {
+  if (childGraphPattern_._graphPatterns.empty()) {
+    auto pattern = std::get<parsedQuery::GroupGraphPattern>(op);
+    childGraphPattern_ = std::move(pattern._child);
+  }
```

Comment:
The else case may definitely not be silent, if this might be set only once, make it an `optional`, and throw an Exception if it is set twice.